[Dependency Cache] Dependency Cache on CI per Project [without GRADLE_RO_DEP_CACHE]

ParaskP7 opened the pull request:
Related: #13122

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This is a test only PR that attempts to test another dependency cache mechanism, one that does **NOT** dependent on the read-only cache ([GRADLE_RO_DEP_CACHE](https://docs.gradle.org/current/userguide/dependency_caching.html#sec:shared-readonly-cache)), and per project.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

`N/A`

> [!TIP]
> To verify that `GRADLE_RO_DEP_CACHE` is **NOT** enabled make sure you don't see this `Shared read-only dependency cache is an incubating feature.` log when Gradle launches for a CI build ([example](https://buildkite.com/automattic/woocommerce-android/builds/26079#0193de2d-dc26-446f-b013-fbc5b6faa7cf/228-239)).

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
CI: Add a scheduled dependency cache job

FYI: This job will be then used by 'buildkite-ci' and configured as a
'buildkite_pipeline_schedule' with a weekly frequency.

ParaskP7 committed:
CI: Use android-staging agent to test with ro dependency cache disabled

AMI ID: ami-06a633c14b35c40b9
AMI Name: android-build-image-6.12.0v1.4-rc-1

ParaskP7 committed:
CI: Remove pre command hook to ignore restore gradle dependency cache

With this hook present, every build and its within jobs are all
restoring the dependency cache, even when this is not really needed, and
as such potentially delaying some jobs unnecessarily, while also
consuming unnecessary resources (downloading from s3, (un)compressing,
etc).

FYI: It is better to use 'restore_gradle_dependency_cache' only on those
jobs that could really benefit from it.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
CI: Temporarily remove all pipeline steps but the mobile app prototype

ParaskP7 committed:
CI: Add restore cache command to run only before mobile app prototype

Since the hook was removed (145a096b6c486eac450041b148e436ae3f57807b),
no build and its within jobs can restore the dependency cache, that is,
unless if it is explicitly specified, just like it is done for mobile
app prototype step via this change.

ParaskP7 committed:
CI: Use non read-only dependency cache to save/restore the cache

ParaskP7 committed:
CI: Revert the removal of all pipeline steps

This change effectively reverts this initial
0dcd7b8016ec60e49268d99ca96ed5c3d55df0ad commit.

ParaskP7 committed:
CI: Cache all dependencies as part of the scheduled dependency cache job

ParaskP7 committed:
CI: Add restore cache command to run on targeted steps

The targeted steps are:
- Mobile App
- Wear App
- Lint
- Detekt
- Unit Tests
- Android tests

ParaskP7 committed:
CI: Check if modules-2 directory exists prior to placing cache

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into ci/enable-dependency-cache-on-ci-v2

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into ci/enable-dependency-cache-on-ci-v2

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into ci/enable-dependency-cache-on-ci-v2

ParaskP7 committed:
CI: Trigger a test build using the latest update from a8c-ci-toolkit

PR: https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/
pull/135

ParaskP7 committed:
CI: Switch back to android agent to re-test without ro dependency cache

After deploying this buildkite-ci#570 change (PR below), which removed
the read-only dependency cache (GRADLE_RO_DEP_CACHE), the agent can be
switched back to 'android' for us to test that the deployment went as
expected, just to be sure.

PR: https://github.com/Automattic/buildkite-ci/pull/570

AMI ID: ami-0fdecc22b38a48860
AMI Name: android-build-image-6.12.0v1.4

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into ci/enable-dependency-cache-on-ci-v2

ParaskP7 committed:
CI: Remove lib:iap unit test as the module is no longer available

For more info see: 14e209ebc380157239af17971546ab06db64ad79

ParaskP7 commented:
Related: #13303

ParaskP7 commented:
Related: #13303
